src/backend/cdb/dispatcher/cdbdisp_query.c

```diff
@@ -1146,9 +1160,19 @@ cdbdisp_dispatchX(QueryDesc* queryDesc,
 		if (si == slice->rootIndex)
 			ds->rootGangSize = primaryGang->size;
 
+		useFullQueryText = queryDesc->extended_query || slice->gangType == GANGTYPE_ENTRYDB_READER;
+		for (int i = 0; i < primaryGang->size; i++)
+			useFullQueryText |= primaryGang->db_descriptors[i]->isWriter;
+
+		if (useFullQueryText)
+			cdbdisp_setDispatchParamsQueryText(ds, queryText, queryTextLength);
+		else
+			cdbdisp_setDispatchParamsQueryText(ds, slimQueryText, slimQueryTextLength);
+
 		if (Test_print_direct_dispatch_info)
-			elog(INFO, "(slice %d) Dispatch command to %s", slice->sliceIndex,
-						segmentsToContentStr(slice->segments));
+			elog(INFO, "(slice %d) Dispatch %s command to %s", slice->sliceIndex,
+			     useFullQueryText ? "full" : "slim",
+			     segmentsToContentStr(slice->segments));
```

Comment:
Why change this?

Reply:
to make the dispatch logic tested. we can always remove this and all test cases changes.

Reply:
Please recover that. Direct dispatch is another thing.